1주차 강의 수강 후 코드 작성

jihwooon opened the pull request:
강의 소스 코드 작성 후 올리기

jihwooon committed:
1주차 강의 수강 후 코드 작성

강의 소스 코드 작성 후 올리기

jihwooon committed:
Feat : 1주차 Todo REST API 과제 피드백

상태 코드 , PORT, BACKLOG 상수 처리 함

- HttpStatus : enum 으로 각 상태 코드 마다 상수 처리
- App : method && path 코드 안에 상태 코드 추가
- DemoHttpHandler : PORT, BACKLOG 숫자 대신 상수 추가

과제물 늦게 올린 점 죄송합니다. 남은 피드백 기간 동안 최대한 열심히 받도록 노력 하겠습니다.!

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,108 @@
+    private ObjectMapper objectMapper = new ObjectMapper();
+    private List<Task> tasks = new ArrayList<>();
+    private HttpStatus httpStatus;
+
+
```
여기 두 줄이 띄어져 있습니다

jihwooon replied:
혹시 두 줄 띄어져 있으면 어떤 의미가 있을까요??

hannut91 replied:
왜 여기만 두 줄이 띄어져 있는지 의문이 들어서요. 여기만 두 줄을 띄우신 의도가 있으실까요?

hannut91 commented on app/src/main/java/com/codesoom/assignment/models/Task.java:
```diff
@@ -0,0 +1,28 @@
+        this.title = title;
+    }
+
+    public String toString() {
+        return "id: " + id + " " +"title: "+ title;
+    }
```
`String.format`을 사용해보는 것은 어떨까요?

hannut91 commented on app/src/main/java/com/codesoom/assignment/InitDto.java:
```diff
@@ -0,0 +1,20 @@
+import com.codesoom.assignment.models.Task;
+
+//따로 DB를 호출 할수 있게 생성 했습니다.
+public class InitDto {
```
`Init`이 의미하는 바가 무엇인가요?

jihwooon replied:
initiation 라는 의미로 쓸려고 했는데 만들고 보니 더 복잡해진거 같습니다.. 다음 리퀘스트 할 때 지울 예정입니다.;;

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,108 @@
+            System.out.println(task);
+        }
+
+    String content = " ";
+
+    HttpStatus httpStatus = HttpStatus.HTTP_NOT_FOUND;
+        //path 부분은 우선 하드 코딩으로 대체 했습니다. String 타입을 쪼개는 방법을 찾으면 변경 할 예정 입니다.
```
여기 코드들의 인덴트가 달라서 읽기 어렵네요. IDE에서 제공하는 자동 정렬 기능을 사용해서 정렬하면 좋을 것 같아요

jihwooon replied:
다시 정렬 해보겠습니다!

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,108 @@
+        //path 부분은 우선 하드 코딩으로 대체 했습니다. String 타입을 쪼개는 방법을 찾으면 변경 할 예정 입니다.
+
+        //path 주소에 따라 content 값이 다르게 나와야 합니다.
+        if ("GET".equals(method) && "/tasks".equals(path)) {
+            content = "[]";
+            httpStatus= HttpStatus.HTTP_OK;
+        }
+
+        if("GET".equals(method) && "/tasks/1".equals(path)) {
+            content = taskToJson();
+            httpStatus= HttpStatus.HTTP_OK;
+        }
```
`else if`를 사용하는 대신 `if`를 사용하신 것 같아요. 이렇게 사용할 경우 여러개의 조건문을 실행할 수 도 있다 라고 해석이 될 것 같아요.

이렇게 사용하는 것보다는 차라리 무조건 조건중에 하나만 실행된다는 `else if`를 사용하거나 혹은 Guard clause방식으로 사용해서 의도를 드러내면 좋을 것 같습니다.

jihwooon replied:
`의도`를 분명히 해야 한다는 말씀 잘 새겨 듣겠습니다~!!

hannut91 commented on app/src/main/java/com/codesoom/assignment/HttpStatus.java:
```diff
@@ -0,0 +1,32 @@
+ * 201 : Create
+ * 204 : No Content
+ * 404 : Not Found
+ *  */
```
코드 마지막 라인에 빈 줄을 추가해 주세요. 혹은 종립님이 하신 것 처럼 자동으로 되도록 설정해도 좋을 것 같아요'

See also
  - https://johngrib.github.io/wiki/intellij/#%ED%8C%8C%EC%9D%BC-%EB%A7%88%EC%A7%80%EB%A7%89%EC%97%90-%EC%9E%90%EB%8F%99%EC%9C%BC%EB%A1%9C-1%EA%B0%9C%EC%9D%98-%EA%B0%9C%ED%96%89-%EB%AC%B8%EC%9E%90-%EC%B6%94%EA%B0%80%ED%95%98%EA%B8%B0

jihwooon committed:
Feat : 1주 Todo REST API 과제 피드백 #2

##개선한 점

- 1차 과제 피드백 개선함
- InitDTO 파일 제거함
- DELETE 부분은 `Transfer-encoding: chunked` 메세지는 나오지만 기능이 아직 구현이 안 됨

## 앞으로 개선 해야 할 부분

- path 부분에서 `{1}`을 구현해야 함
- 생성자를 만들어서 List 에 데이터를 셋팅하는 방법이 아닌 다른 방법으로 구현 해야 함
- POST 로 저장된 값을 PUT과 PATCH 를 통해 값을 바꿀 수 있어야 함

jihwooon commented on app/src/main/java/com/codesoom/assignment/models/Task.java:
```diff
@@ -23,6 +23,8 @@ public void setTitle(String title) {
 
     public String toString() {
-        return "id: " + id + " " +"title: "+ title;
+        String id = "id";
```
String id와  String title 가 아니라 this.id = id, this.title = title 입니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/models/Task.java:
```diff
@@ -0,0 +1,30 @@
+        this.title = title;
+    }
+
+    public String toString() {
+        String id = "id";
+        String title = "title";
+        return String.format("Task-> id: %s, title: %s", id, title);
+    }
```
아직 작업중이신 것 같긴한데, `String.format("Task-> id: %s, title: %s", this.id, this.title);`을 의도하신 것 맞나요?

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    private String taskToJson() throws IOException {
+
```
다른 곳에는 메서드 바로 아래에 빈 줄이 없는데 여기에는 빈 줄이 있습니다. 다른 의도가 있으신가요?

jihwooon replied:
아마 제가 코드 치면서 엔터는 치는 습관을 가지고 있어서 중간에 빈 공간이 생기는 거 같습니다. 의도 라기 보다는 습관에 문제라서 최대한 이 부분을 고치도록 노력 해보겠습니다~!

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,107 @@
+        //path 부분은 우선 하드 코딩으로 대체 했습니다. String 타입을 쪼개는 방법을 찾으면 변경 할 예정 입니다.
+
+        //path 주소에 따라 content 값이 다르게 나와야 합니다.
+        if ("GET".equals(method) && "/tasks".equals(path)) {
+            content = taskToJson();
+            httpStatus = HttpStatus.HTTP_OK;
+        }
+        else if
+            ("GET".equals(method) && "/tasks/1".equals(path)) {
+            content = taskToJson();
+            httpStatus = HttpStatus.HTTP_OK;
+        };
```
```suggestion
        if ("GET".equals(method) && "/tasks".equals(path)) {
            content = taskToJson();
            httpStatus = HttpStatus.HTTP_OK;
        } else if ("GET".equals(method) && "/tasks/1".equals(path)) {
            content = taskToJson();(
            httpStatus = HttpStatus.HTTP_OK;
        };
```

공식이 있는 것은 아니지만 저는 이게 더 읽기 좋은 것 같아요

jihwooon replied:
저도 사실 밑에 부분이 읽기 더 편하다고 생각합니다. ^^;;

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,107 @@
+
+        String content = " ";
+        HttpStatus httpStatus = HttpStatus.HTTP_NOT_FOUND;
+        //path 부분은 우선 하드 코딩으로 대체 했습니다. String 타입을 쪼개는 방법을 찾으면 변경 할 예정 입니다.
+
+        //path 주소에 따라 content 값이 다르게 나와야 합니다.
```
코드를 읽는 사람을 위하여 주석을 남겨놓으신 것 같아요. 주석보다는 GitHub에 지금과 같이 코멘트를 남기는게 더 좋을 것 같습니다

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,107 @@
+        }
+
+        String content = " ";
+        HttpStatus httpStatus = HttpStatus.HTTP_NOT_FOUND;
```
기본으로 Not Found로 설정을 해놓고, 이 변수를 변경해서 응답을 하거나 혹은 응답할 적절한 요청을 찾지 못한다면 Not Found를 그대로 반환하기 위해서 기본 값으로 설정해놓으신 것 같아요.

이러한 변수는 코드를 읽는 사람에게 있어서 복잡성을 높이게 됩니다. 왜냐하면 코드를 읽으면서 이 변수가 변할 것이라는 것을 인지한 상태에서 계속해서 코드를 읽어나가야 하기 때문입니다.

그렇다면 어떻게 문제를 해결하는 것이 좋을까요?

jihwooon replied:
이 부분을 항상 염두하면서 계속해서 답을 찾아 보도록 해보겠습니다~!

hannut91 commented on app/src/main/java/com/codesoom/assignment/App.java:
```diff
@@ -1,10 +1,34 @@
         return "Hello World!";
     }
 
+    // 상수 일 경우 변수명을 대문자로 표기 해야한다
+    final static int PORT = 8000;
+    final static int BACKLOG = 0;
```
상수의 경우에는 클래스 선언 바로 아래에 위치하고, 그 다음에 메서드들이 위치하는게 읽을 때 더 좋을 것 같아요

hannut91 commented on app/src/main/java/com/codesoom/assignment/DemoHttpHandler.java:
```diff
@@ -0,0 +1,107 @@
+        String method = exchange.getRequestMethod();
+        URI uri = exchange.getRequestURI();
+        String path = uri.getPath();
+        InputStream inputStream = exchange.getRequestBody();
+
+        String body = new BufferedReader(new InputStreamReader(inputStream)).lines().collect(Collectors.joining("\n"));
+
+        System.out.println(method + "" + path);
+
+        if (!body.isBlank()) {
+            System.out.println(body);
+
+            Task task = toTask(body);
+            tasks.add(task);
+
+            System.out.println(task);
+        }
```
우리가 글을 작성할 때 관련이 있는 것과 관련이 없는 것을 문단으로 구분하는 것처럼, 코드도 동일하게 적절하게 잘 묶어주면
코드를 읽는 사람이 더 잘 읽을 수 있습니다.

```suggestion
        InputStream inputStream = exchange.getRequestBody();
        String body = new BufferedReader(
            new InputStreamReader(inputStream)
        ).lines().collect(Collectors.joining("\n"));
        if (!body.isBlank()) {
            Task task = toTask(body);
            tasks.add(task);
        }
```

hannut91 commented:
고생 많으셨습니다 ㅎㅎ 답을 하지 못하신 부분은 계속해서 다루는 주제이니까 다음 과제에서 더 시도해보실 수 있으실 거에요. 간략하게 말씀드리면 값이 변하는 변수 대신에 immutable한 변수 즉 상수를 사용하는게 코드의 복잡성을 낮출 수 있습니다.

고생 많으셨습니다!

hannut91 commented:
아 그리고 풀 리퀘스트는 `CodeSoom:main`이 아니라 `CodeSoom:jihwooon`으로 보내주셔야 합니다. 

![image](https://user-images.githubusercontent.com/14071105/144752019-cdcc1640-da8d-4951-860a-fcbef7526c1c.png)

Edit을 누른 후 

![image](https://user-images.githubusercontent.com/14071105/144752033-f4bd5498-d78f-482b-b346-10f1c312fd80.png)

위 버튼을 눌러 변경할 수 있습니다.

변경해 주시면 머지 하겠습니다!
